File: Assets/Script/Tools/TexturePacker/Packer.lua
-- [ts]: Packer.ts
local ____exports = {} -- 1
local function CreatePacker() -- 39
	local packer = { -- 40
		fit = function(self, blocks) -- 41
			table.sort( -- 42
				blocks, -- 42
				function(a, b) -- 42
					return math.max(a.w, a.h) > math.max(b.w, b.h) -- 43
				end -- 42
			) -- 42
			local len = #blocks -- 45
			local w = len > 0 and blocks[1].w or 0 -- 46
			local h = len > 0 and blocks[1].h or 0 -- 47
			self.root = {x = 0, y = 0, w = w, h = h} -- 48
			for ____, block in ipairs(blocks) do -- 49
				local node = self:findNode(self.root, block.w, block.h) -- 50
				if node then -- 50
					block.fit = self:splitNode(node, block.w, block.h) -- 52
				else -- 52
					block.fit = self:growNode(block.w, block.h) -- 54
				end -- 54
			end -- 54
		end, -- 41
		findNode = function(self, node, w, h) -- 59
			if node.used then -- 59
				return node.right and self:findNode(node.right, w, h) or node.down and self:findNode(node.down, w, h) -- 61
			elseif w <= node.w and h <= node.h then -- 61
				return node -- 64
			else -- 64
				return nil -- 66
			end -- 66
		end, -- 59
		splitNode = function(self, node, w, h) -- 70
			node.used = true -- 71
			node.down = {x = node.x, y = node.y + h, w = node.w, h = node.h - h} -- 72
			node.right = {x = node.x + w, y = node.y, w = node.w - w, h = h} -- 78
			return node -- 84
		end, -- 70
		growNode = function(self, w, h) -- 87
			if self.root == nil then -- 87
				return nil -- 88
			end -- 88
			local canGrowDown = w <= self.root.w -- 89
			local canGrowRight = h <= self.root.h -- 90
			local shouldGrowRight = canGrowRight and self.root.h >= self.root.w + w -- 91
			local shouldGrowDown = canGrowDown and self.root.w >= self.root.h + h -- 92
			if shouldGrowRight then -- 92
				return self:growRight(w, h) -- 94
			elseif shouldGrowDown then -- 94
				return self:growDown(w, h) -- 96
			elseif canGrowRight then -- 96
				return self:growRight(w, h) -- 98
			elseif canGrowDown then -- 98
				return self:growDown(w, h) -- 100
			else -- 100
				return nil -- 102
			end -- 102
		end, -- 87
		growRight = function(self, w, h) -- 106
			if self.root == nil then -- 106
				return nil -- 107
			end -- 107
			self.root = { -- 108
				used = true, -- 109
				x = 0, -- 110
				y = 0, -- 111
				w = self.root.w + w, -- 112
				h = self.root.h, -- 113
				down = self.root, -- 114
				right = {x = self.root.w, y = 0, w = w, h = self.root.h} -- 115
			} -- 115
			local node = self:findNode(self.root, w, h) -- 122
			if node then -- 122
				return self:splitNode(node, w, h) -- 124
			else -- 124
				return nil -- 126
			end -- 126
		end, -- 106
		growDown = function(self, w, h) -- 130
			if self.root == nil then -- 130
				return nil -- 131
			end -- 131
			self.root = { -- 132
				used = true, -- 133
				x = 0, -- 134
				y = 0, -- 135
				w = self.root.w, -- 136
				h = self.root.h + h, -- 137
				down = {x = 0, y = self.root.h, w = self.root.w, h = h}, -- 138
				right = self.root -- 144
			} -- 144
			local node = self:findNode(self.root, w, h) -- 146
			if node then -- 146
				return self:splitNode(node, w, h) -- 148
			else -- 148
				return nil -- 150
			end -- 150
		end -- 130
	} -- 130
	return packer -- 154
end -- 39
____exports.default = CreatePacker -- 157
return ____exports -- 157

File: Assets/Script/Tools/TexturePacker/Packer.ts
/* Copyright (c) 2016-2025 Li Jin <dragon-fly@qq.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

import { Sprite } from 'Dora';

export interface Block {
	name: string;
	w: number;
	h: number;
	sp?: Sprite.Type;
	fit?: Node;
}

export interface Node {
	x: number;
	y: number;
	w: number;
	h: number;
	right?: Node;
	down?: Node;
	used?: boolean;
}

interface Packer {
	root?: Node;
	fit(blocks: Block[]): void;
	findNode(node: Node, w: number, h: number): Node | undefined;
	splitNode(node: Node, w: number, h: number): Node;
	growNode(w: number, h: number): Node | undefined;
	growRight(w: number, h: number): Node | undefined;
	growDown(w: number, h: number): Node | undefined;
}

function CreatePacker() {
	const packer: Packer = {
		fit(blocks: Block[]) {
			table.sort(blocks, (a, b) => {
				return math.max(a.w, a.h) > math.max(b.w, b.h);
			});
			const len = blocks.length;
			const w = len > 0 ? blocks[0].w : 0;
			const h = len > 0 ? blocks[0].h : 0;
			this.root = {x: 0, y: 0, w, h};
			for (let block of blocks) {
				const node = this.findNode(this.root, block.w, block.h);
				if (node) {
					block.fit = this.splitNode(node, block.w, block.h);
				} else {
					block.fit = this.growNode(block.w, block.h);
				}
			}
		},

		findNode(node: Node, w: number, h: number): Node | undefined {
			if (node.used) {
				return (node.right && this.findNode(node.right, w, h))
					|| (node.down && this.findNode(node.down, w, h));
			} else if (w <= node.w && h <= node.h) {
				return node;
			} else {
				return undefined;
			}
		},

		splitNode(node: Node, w: number, h: number): Node {
			node.used = true;
			node.down = {
				x: node.x,
				y: node.y + h,
				w: node.w,
				h: node.h - h
			};
			node.right = {
				x: node.x + w,
				y: node.y,
				w: node.w - w,
				h
			};
			return node;
		},

		growNode(w: number, h: number): Node | undefined {
			if (this.root === undefined) return undefined;
			const canGrowDown = w <= this.root.w;
			const canGrowRight = h <= this.root.h;
			const shouldGrowRight = canGrowRight && (this.root.h >= this.root.w + w);
			const shouldGrowDown = canGrowDown && (this.root.w >= this.root.h + h);
			if (shouldGrowRight) {
				return this.growRight(w, h);
			} else if (shouldGrowDown) {
				return this.growDown(w, h);
			} else if (canGrowRight) {
				return this.growRight(w, h);
			} else if (canGrowDown) {
				return this.growDown(w, h);
			} else {
				return undefined;
			}
		},

		growRight(w: number, h: number): Node | undefined {
			if (this.root === undefined) return undefined;
			this.root = {
				used: true,
				x: 0,
				y: 0,
				w: this.root.w + w,
				h: this.root.h,
				down: this.root,
				right: {
					x: this.root.w,
					y: 0,
					w,
					h: this.root.h
				}
			};
			const node = this.findNode(this.root, w, h);
			if (node) {
				return this.splitNode(node, w, h);
			} else {
				return undefined;
			}
		},

		growDown(w: number, h: number): Node | undefined {
			if (this.root === undefined) return undefined;
			this.root = {
				used: true,
				x: 0,
				y: 0,
				w: this.root.w,
				h: this.root.h + h,
				down: {
					x: 0,
					y: this.root.h,
					w: this.root.w,
					h
				},
				right: this.root
			};
			const node = this.findNode(this.root, w, h);
			if (node) {
				return this.splitNode(node, w, h);
			} else {
				return undefined;
			}
		}
	};
	return packer;
}

export default CreatePacker;
